sled-storage/src/manager.rs

```diff
@@ -590,9 +663,145 @@ impl StorageManager {
             info!(self.log, "KeyManager ready, but no ledger detected");
         }
 
+        // We don't load any configuration for datasets, since we aren't
+        // currently storing any dataset information in-memory.
+        //
+        // If we ever wanted to do so, however, we could load that information
+        // here.
+
         Ok(())
     }
 
+    async fn datasets_ensure(
+        &mut self,
+        mut config: DatasetsConfig,
+    ) -> Result<DatasetsManagementResult, Error> {
+        let log = self.log.new(o!("request" => "datasets_ensure"));
+
+        // Ensure that the datasets arrive in a consistent order
+        config.datasets.sort_by(|a, b| a.id.partial_cmp(&b.id).unwrap());
+
+        // We rely on the schema being stable across reboots -- observe
+        // "test_datasets_schema" below for that property guarantee.
+        let ledger_paths = self.all_omicron_dataset_ledgers().await;
+        let maybe_ledger =
+            Ledger::<DatasetsConfig>::new(&log, ledger_paths.clone()).await;
+
+        let mut ledger = match maybe_ledger {
+            Some(ledger) => {
+                info!(
+                    log,
+                    "Comparing 'requested datasets' to ledger on internal storage"
+                );
+                let ledger_data = ledger.data();
+                if config.generation < ledger_data.generation {
+                    warn!(
+                        log,
+                        "Request looks out-of-date compared to prior request"
+                    );
+                    return Err(Error::DatasetConfigurationOutdated {
+                        requested: config.generation,
+                        current: ledger_data.generation,
+                    });
+                } else if config.generation == ledger_data.generation {
+                    info!(
+                        log,
+                        "Requested generation number matches prior request",
+                    );
+
+                    if ledger_data != &config {
+                        error!(
+                            log,
+                            "Requested configuration changed (with the same generation)";
+                            "generation" => ?config.generation
+                        );
+                        return Err(Error::DatasetConfigurationChanged {
+                            generation: config.generation,
+                        });
+                    }
+                } else {
+                    info!(log, "Request looks newer than prior requests");
+                }
+                ledger
+            }
+            None => {
+                info!(log, "No previously-stored 'requested datasets', creating new ledger");
+                Ledger::<DatasetsConfig>::new_with(
+                    &log,
+                    ledger_paths.clone(),
+                    DatasetsConfig::default(),
```

Comment:
This has a generation of 1, right? Does the input in `config` always have a generation >= 2, or do we care about that?

Reply:
In this case, we're going to take the ledger data and set it to the user-supplied config anyway (see: `*ledger_data = config` below).

So, the user can supply whatever generation number they want if the ledger hasn't seen any prior requests. However, on all *subsequent* requests, we'll make sure the generation number is increasing.